identity/aziot-identityd/src/lib.rs

```diff
@@ -702,73 +622,164 @@ impl Api {
+                        Error::Internal(InternalError::CreateCertificate(Box::new(err)))
+                    })?;
+                let certificate = self
+                    .cert_client
+                    .create_cert(&module_id, &csr, None)
+                    .await
+                    .map_err(|err| {
+                        Error::Internal(InternalError::CreateCertificate(Box::new(err)))
+                    })?;
+                let certificate = String::from_utf8(certificate).map_err(|err| {
+                    Error::Internal(InternalError::CreateCertificate(Box::new(err)))
+                })?;
+
+                // Parse certificate expiration time.
+                let expiration = get_cert_expiration(&certificate)?;
+
+                aziot_identity_common::Identity::Local(aziot_identity_common::LocalIdSpec {
+                    module_id: module_id.to_owned(),
+                    auth: aziot_identity_common::LocalAuthenticationInfo {
+                        private_key: private_key_pem,
+                        certificate,
+                        expiration,
+                    },
+                })
+            }
+        };
+
+        Ok(local_identity)
+    }
+
     async fn create_identity_cert_if_not_exist_or_expired(
         &self,
         identity_pk: &str,
         identity_cert: &str,
         subject: &str,
-        attributes: Option<aziot_identity_common::LocalIdAttr>,
     ) -> Result<(), Error> {
-        let device_id_cert = self.cert_client.get_cert(identity_cert).await;
-        let create_cert = match device_id_cert {
-            Ok(device_id_cert) => {
-                let x509_req = openssl::x509::X509::from_pem(&device_id_cert).map_err(|err| {
+        // Retrieve existing cert and check it for expiry.
+        let device_id_cert = match self.cert_client.get_cert(identity_cert).await {
+            Ok(pem) => {
+                let cert = openssl::x509::X509::from_pem(&pem).map_err(|err| {
                     Error::Internal(InternalError::CreateCertificate(Box::new(err)))
                 })?;
-                let cert_expiration = x509_req.as_ref().not_after();
+                let cert_expiration = cert.as_ref().not_after();
                 let current_time = openssl::asn1::Asn1Time::days_from_now(0).map_err(|err| {
                     Error::Internal(InternalError::CreateCertificate(Box::new(err)))
                 })?;
                 let expiration_time = current_time.diff(cert_expiration).map_err(|err| {
                     Error::Internal(InternalError::CreateCertificate(Box::new(err)))
                 })?;
-                expiration_time.days < 1
+
+                if expiration_time.days < 1 {
+                    log::info!("{} has expired. Renewing certificate", identity_cert);
+
+                    None
+                } else {
+                    Some(pem)
+                }
             }
-            Err(_) => true,
+            Err(_) => None,
```

Comment:
We need to check the type of error. If the CS was not accessible, we should not try and delete the preloaded key.

Reply:
Yes. Though since we don't actually delete the key right now, I just left it as that.

Reply:
Please add a note in the TODO comment to ensure that case doesn't get missed when we implement the delete key. Thanks!

Reply:
Added